### PR TITLE
test: append to corrupted journals

### PR DESCRIPTION
**Pull request by mrc0mmand**
Introduce a manual test tool that creates a journal, corrupts it by
flipping bits at given offsets, and then attempts to write to the journal.
In ideal case we should handle this gracefully without any crash or
memory corruption.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/journal/test-journal-append.c [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+static int journal_corrupt_and_append(uint64_t start_offset, uint64_t step) {
+        _cleanup_(mmap_cache_unrefp) MMapCache *mmap_cache = NULL;
+        _cleanup_(rm_rf_physical_and_freep) char *tempdir = NULL;
+        _cleanup_(managed_journal_file_closep) ManagedJournalFile *mj = NULL;
+        uint64_t start, end;
+        int r;
+
+        mmap_cache = mmap_cache_new();
+        assert_se(mmap_cache != NULL);
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: `!= NULL` is not necessary.

File: src/journal/test-journal-append.c [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                return log_error_errno(r, "Failed to open the journal: %m");
+
+        assert_se(mj);
+        assert_se(mj->file);
+
+        /* Add a couple of initial messages */
+        for (int i = 0; i < 10; i++) {
+                _cleanup_free_ char *message = NULL;
+
+                assert_se(asprintf(&message, "MESSAGE=Initial message %d", i));
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: `assert_se(asprintf(...) >= 0);`

File: src/journal/test-journal-append.c [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+                r = journal_append_message(mj, message);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write to the journal: %m");
+        }
+
+        start = start_offset == UINT64_MAX ? random_u32() % mj->file->last_stat.st_size : start_offset;
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: Not a matter, but why not `random_u64()` ?

File: src/journal/test-journal-append.c [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                if (r < 0) {
+                        /* The corrupted journal might get rejected during reopening
+                         * if it's corrupted enough (especially its header), so
+                         * treat this as a success if it doesn't crash */
+                        log_info_errno(r, "Failed to reopen the journal: %m");
+                        break;
+                }
+
+                /* Try to write something to the (possibly corrupted) journal */
+                assert_se(asprintf(&message, "MESSAGE=Hello world %" PRIu64, offset));
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: missing `>= 0`

File: src/journal/test-journal-append.c [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        return 0;
+}
+
+int main(int argc, char *argv[]) {
+        uint64_t start_offset = UINT64_MAX;
+        uint64_t iterations = 100;
+        uint64_t iteration_step = 1;
+        uint64_t corrupt_step = 31;
+        bool sequential = false;
+        bool run_one = false;
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: Not a matter, but better to merge lines?

File: src/journal/test-journal-append.c [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                { "corrupt-step",        required_argument, NULL, ARG_CORRUPT_STEP        },
+                { "sequential",          no_argument,       NULL, ARG_SEQUENTIAL          },
+                { "run-one",             required_argument, NULL, ARG_RUN_ONE             },
+                {}
+        };
+
+        assert_se(argc >= 0);
+        assert_se(argv);
+
+        while ((c = getopt_long(argc, argv, "", options, NULL)) >= 0)
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: `""` -> `"h"`

File: src/journal/test-journal-append.c [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+                               "                            (default: 100)\n"
+                               "    --iteration-step=STEP   Iteration step (default: 1)\n"
+                               "    --corrupt-step=STEP     Corrupt every n-th byte starting from OFFSET (default: 31)\n"
+                               "    --sequential            Go through offsets sequentially instead of picking\n"
+                               "                            a random one on each iteration. If set, we go through\n"
+                               "                            offsets <0; ITER), or <OFFSET, ITER) if --start-offset=\n"
+                               "                            is set (default: false)\n"
+                               "    --run-one=OFFSET        Single shot mode for reproducing issues. Takes the same\n"
+                               "                            offset as --start-offset= and does only one iteration\n"
+                               , argv[0]);
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: `program_invocation_short_name` is better?

File: src/journal/test-journal-append.c [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+                /* Reproducer mode */
+                return journal_corrupt_and_append(start_offset, corrupt_step);
+
+        for (uint64_t i = 0; i < iterations; i++) {
+                uint64_t offset = UINT64_MAX;
+
+                log_info("Iteration #%" PRIu64 ", step: %" PRIu64, i, iteration_step);
+
+                if (sequential)
+                    offset = (start_offset == UINT64_MAX ? 0 : start_offset) + i * iteration_step;
```
Comment (yuwata) [inline src/journal/test-journal-append.c]: Indentation.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/journal/test-journal-append.c (hunk 3 above)
Comment (mrc0mmand) [reply src/journal/test-journal-append.c]: No idea to be honest :)

File: src/journal/test-journal-append.c (hunk 7 above)
Comment (mrc0mmand) [reply src/journal/test-journal-append.c]: Ah, interesting, didn't know about this one.

**Review by yuwata — APPROVED** [review]

**Pull request merged**